### DISCUSSION TITLE
Port Babel plugin to SWC for Next.js

### DISCUSSION THREAD

**Discussion by nandorojo** [software-mansion/react-native-reanimated#2586] [Ideas]
The new version of Next.js uses SWC instead of Babel. This results in way faster bundle and startup times. 

The one reason I can't use this is because I need the Reanimated 2 babel plugin. Is there any shot someone would know how to port this to SWC? I have no idea how either babel or SWC work, but I would love to get to take advantage of Next.js 12's innovations...

For context, I just gave a talk about Next.js + React Native: https://twitter.com/FernandoTheRojo/status/1453712622222729217?s=20

Relevant Next.js thread & comment: https://github.com/vercel/next.js/discussions/30174#discussioncomment-1540972

Any guidance would be great, thanks!

## Update

I opened a new discussion about disabling the Babel plugin altogether for Web: https://github.com/software-mansion/react-native-reanimated/discussions/2620

**Comment by wibb36** [comment]
Same here. You can always still use Babel (this will disable SWC) with Next 12 while we figure this out.

**Reply by nandorojo** [reply]
Yeah, this is what I’m doing in the interim. It would be awesome to take advantage of the great new Next.js architecture though.

**Reply by wibb36** [reply]
Absolutely agree. For kicks, are you facing any other issues using Next.js 12 and Babel? 

regarding SWC, I think there should be a reanimated plugin for SWC. I'll keep you updated

**Reply by nandorojo** [reply]
So far, I think it’s working. Reanimated 2.2.4 I think doesn’t work with Next.js (I get `component.previousStyle` warnings) but overall I’ve been fine with Next.js 12 and Babel.

Is the SWC + Reanimated plug-in something you would pick up?

**Reply by wibb36** [reply]
I'm not sure if it's possible for me at the moment because:

1. I don't speak RUST and it seems to be the only way so far to creating a plugin, consequently, recreating the reanimated plugin. There's unresolved parsing issues with JS.
2. I'm not 100% sure if porting reanimated babel plugin to SWC is the only approach.
3. As you already know, an ideal scene would be running animations as natively as possible cross-platform, a departure from that ideal in this case would be running CSS instead of JS on the web therefore, maybe it is worth focusing on making reanimated / maybe MOTI :) running CSS for the web. I'm not sure if you agree with me on this last one??

PS: I think, the good news is that SWC supports all the ecmascript needed in case someone here rewrite the plugin in Rust.

**Comment by nandorojo** [comment]
I opened a new discussion about disabling the Babel plugin altogether for Web: https://github.com/software-mansion/react-native-reanimated/discussions/2620

**Comment by nandorojo** [comment]
If someone who is very familiar with Reanimated could give me a sense of how the babel plugin works, I can try to relay this to the Next.js team to see if they can port it.

**Comment by axeldelafosse** [comment]
Looking for a Rustacean to work on this: https://twitter.com/axeldelafosse/status/1493992830175481859

**Comment by piaskowyk** [comment]
Interesting thread! If someone wants to port reanimated plugin to SWC I can help with pleasure. For web is required just a part of the reanimated plugin.

**Reply by nandorojo** [reply]
that would be amazing! the port would only be necessary for Web, to add Next.js compatibility. @axeldelafosse

**Reply by axeldelafosse** [reply]
Woohoo! Good news, thank you! Could you highlight what part of the plugin is required for web please?

**Reply by wibb36** [reply]
Can't wait. Fernando @nandorojo  , you mentioned in one of the discussions that you have a patch for reanimated?? 

I'd be happy to contribute with $$$ if that helps.

**Reply by nandorojo** [reply]
I tried, but no patch. I think if you use dependency arrays for all reanimated hooks, you’ll do most of what the babel plug-in does. But I would need info from someone more knowledgeable about what the plug-in actually does.

**Reply by wibb36** [reply]
Make perfect sense. @piaskowyk How do you plan to help?

**Reply by piaskowyk** [reply]
Sorry for my absence, I had to deal with my other duties at work. I will try to describe everything on Monday.

**Reply by piaskowyk** [reply]
The plugin is needed for the web just to generate closure. 
`const inputs = Object.values(updater._closure ?? {});` - 
https://github.com/software-mansion/react-native-reanimated/blob/main/src/reanimated2/hook/useAnimatedStyle.ts#L412
`_closure` - https://github.com/software-mansion/react-native-reanimated/blob/main/plugin.js#L462-L463

But you can simplify it and insert to `_closure` just SharedValues used by updater function passed as an argument to useAnimatedStyle (`useAnimatedStyle(updater)`). It should be enough.

**Reply by nandorojo** [reply]
> But you can simplify it and insert to _closure just SharedValues used by updater function passed as an argument to useAnimatedStyle (useAnimatedStyle(updater)). It should be enough.

I assume we would want non-shared values as well, right? To make sure it’s reactive.

**Reply by piaskowyk** [reply]
You are right. The `_closure` is added as a dependency, but if you want you can specific dependency array as the second argument for `useAnimatedStyle`. In other words, If your plugin won't prepare the whole `_closure` it will work but you'll specify the dependency array manually.

**Reply by nandorojo** [reply]
@piaskowyk do you know of anyone who might be able to help make the rust plug-in? I think there are enough people who would be willing to sponsor if so!

happy to jump on a quick call to discuss that if that's easier

**Reply by axeldelafosse** [reply]
Yeah let's chat together (Krzysztof, Fernando and I) to discuss the next steps. I'll send you an email

**Comment by jesse-savary** [comment]
Would be great to see this; wondering how well it would pair with https://github.com/callstack/repack/ for Android/iOS

**Comment by bran-dy-b** [comment]
Was there any conclusion here with respect to whether or not this is going to happen? I saw this but there doesn't seem to be any activity around it https://github.com/software-mansion/react-native-reanimated/pull/3341

Would be great to be able to use SWC and drop babel, but I don't have the expertise to contribute here 😞

**Reply by axeldelafosse** [reply]
Hey Brandon! We sponsored this open-source work and Jakub from Software Mansion released it here: https://github.com/j-piasecki/swc-plugin-react-native-reanimated. You can refer to the Showtime Next.js app for usage: https://github.com/showtime-xyz/showtime-frontend/blob/staging/apps/next/next.config.js#L92
